### PR TITLE
Cache conda environment between CI test runs

### LINKED ISSUE #6853: Setting up conda env takes 2-8min per CI job — cache it?
Looking at a [recent CI run](https://github.com/dask/distributed/runs/7724401999?check_suite_focus=true) on main, I see the `Setup Conda Environment` step taking 2-3min on linux, and typically 6-7min on Windows and macOS. That's 10-15% of total job runtime.

Could we cache the conda environment between CI runs, and only reinstall when the `environment.yaml` changes? (And maybe once per day too, so things don't get too stale?)

A downside is that most of our CI dependencies are un-pinned, which we're relying on as a crude form of upstream testing: when dependencies release new versions, our CI might non-deterministically go red if they broke something. Personally, I think this is a bad experience for contributors and bad practice for CI in general, but it doesn't seem like we're going to change this right now https://github.com/dask/community/issues/203. So if we did cache conda envs, it might add more latency when upstream dependencies break us.

cc @fjetter @quasiben

### PR DESCRIPTION
**Pull request by charlesbluca**
Closes #6853 

Follows the [steps](https://github.com/conda-incubator/setup-miniconda#caching-environments) outlined in the `setup-miniconda` docs to cache the conda environment between test runs.

Note that (as pointed out by @gjoseph92), since we use unpinned CI dependencies in our environments, it's possible for the cached environment to become outdated - hoping to address this with a triggering keyword that can be added to PRs or commits to skip the caching and install the environment from scratch.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by gjoseph92** [comment]
Sweet, thanks @charlesbluca!

How can we test this out? Seems like tests don't actually run when you just modify the GitHub actions yamls. Maybe push a spurious change to something else to trigger tests, wait a bit, then push another change and make sure the second build used the cache?

**Comment by charlesbluca** [comment]
I don't see anything in the modified workflow implying that it only runs on changes to certain files, so think we might just be blocked up by other running jobs in the Dask org; I'll ping this PR with an empty commit in a couple hours to see if things run then, but my plan was to do what you described, and then push a commit containing `[skip-caching]` to see if we're able to disable caching as expected 🙂

**Comment by gjoseph92** [comment]
In other PRs, I'm used to seeing jobs at least queued, if not running, as soon as something gets pushed. Like on https://github.com/dask/distributed/pull/6856 right now, I see: 
image

Maybe try pushing a spurious commit right now? Not sure if anything will change if we wait.

**Review by jrbourbeau — COMMENTED** [review]
I'm seeing 

```
Invalid workflow file: .github/workflows/tests.yaml#L49
You have an error in your yaml syntax on line 49
```

over in https://github.com/dask/distributed/actions/runs/2821228837

**Comment by charlesbluca** [comment]
Thanks @jrbourbeau 🤦🏼‍♂️ didn't realize the failed workflow syntax wouldn't pop up here, will fix now

**Comment by charlesbluca** [comment]
Looks like that got things working 🎉

**Comment by gjoseph92** [comment]
Weird, the cache didn't work between those two commits, even though they seemed to have the same cache key. I had figured the cache would be saved as soon as the cache task completed, but maybe because the first CI job was interrupted, it wasn't?

**Comment by charlesbluca** [comment]
It looks like maybe the cache isn't being saved for failed runs?

Unfortunately, due to GitHub Actions running on UTC, the cached environments for the non-macOS runs for https://github.com/dask/distributed/pull/6855/commits/a08262380d8064a66b87991c8f2cd37e1e2a0bc3 are no longer valid and need to be solved again. However, the macOS runs _should_ have created a cached environment for today, but I notice the failed 3.8 / 3.9 runs didn't bother to cache the environment after failing ([example](https://github.com/dask/distributed/runs/7735278192?check_suite_focus=true)).

However, the macOS 3.10 run succeeded and did end up saving the conda environment in the cache, which we can see getting picked up [here](https://github.com/dask/distributed/runs/7737976448?check_suite_focus=true) and [here](https://github.com/dask/distributed/runs/7737976506?check_suite_focus=true). I notice that the time it takes to restore the cache is variable (in one run it takes >30 seconds, in another it takes nearly 4 minutes).

**Comment by charlesbluca** [comment]
Looks like all the OS / python version environments are now cached and we can see the speed up in:

https://github.com/dask/distributed/actions/runs/2822501538

**Comment by charlesbluca** [comment]
Looks like the option to skip caching is working fine 🙂 commits with `[skip-caching]` in the message should now skip over the caching process and resolve the environment, which should be good in cases where we need to pull in the latest CI deps.

Is there anywhere it would make sense to document this option, or the caching behavior?

**Comment by charlesbluca** [comment]
cc @gjoseph92 @jrbourbeau this should be ready for review now

**Review by jacobtomlinson — APPROVED** [review]
This looks great thanks for this @charlesbluca

**Review by gjoseph92 — APPROVED** [review]
Thanks @charlesbluca! I'm going to merge this unless anyone objects?

**Pull request merged**
